### ISSUE TITLE
PoissonVI Error in Pre-release

### ISSUE THREAD

**Issue by shahrozeabbas** [scverse/scvi-tools#2370]
Hello,

I recently posted about using `PoissonVI` and was pointed to the pre-release verison using `pip`. I am currently running that version and have run into an error that I suspect may be from the software. The error is not raised right away but after a few training epochs which is strange. 

```
ValueError: Expected parameter loc (Tensor of shape (128, 16)) of distribution Normal(loc: torch.Size([128, 16]), scale: torch.Size([128, 16])) to satisfy the constraint Real(), but found invalid values:
tensor([[nan, nan, nan,  ..., nan, nan, nan],
        [nan, nan, nan,  ..., nan, nan, nan],
        [nan, nan, nan,  ..., nan, nan, nan],
        ...,
        [nan, nan, nan,  ..., nan, nan, nan],
        [nan, nan, nan,  ..., nan, nan, nan],
        [nan, nan, nan,  ..., nan, nan, nan]], device='cuda:0',
       grad_fn=<AddmmBackward0>)
```

**Comment by shahrozeabbas** [comment]
Quick update: I have updated some of my `jax` libraries and tried on a subset of the full data. It seems to be working fine now, but I will try with the entire dataset and close the issue if the problem is no longer occurring.

**Comment by shahrozeabbas** [comment]
Still having issues with this it seems, please let me know if there's further information I need to provide to solve this, thanks!

**Comment by martinkim0** [comment]
Hi @shahrozeabbas, sorry you're experiencing this issue. Could you try passing in `POISSONVI(var_activation=torch.nn.functional.softplus)` and see if that fixes it? ccing @lauradmartens

**Comment by lauradmartens** [comment]
Hi @shahrozeabbas! Indeed sorry that you are experiencing issues. To try to reproduce your error, I reinstalled the pre-release sci-tools in a new conda environment and ran the model on the data from the tutorial but everything works fine there. Which dataset are you using? Did you try to run the tutorial? Does that work for you?

**Comment by shahrozeabbas** [comment]
@lauradmartens Everything works fine from the tutorial as well as when I try with a single dataset or just a few datasets merged together. I am using unpublished dataset so can't point to anything for you to reproduce unfortunately. 

The error is occuring once I feed in all ~200 samples. I am pasting the whole error message below to provide some insight. The weird thing is that it fails after 6% of the training, so it's not right away, but it runs for a bit then fails. 

```
File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/scripts/tmp90vs87w0.atac_model.py", line 21, in <module>
    model.train(accelerator='gpu', max_epochs=1000, devices=1, plan_kwargs={'lr_factor': 0.1})
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/scvi/model/_peakvi.py", line 241, in train
    super().train(
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/scvi/model/base/_training_mixin.py", line 100, in train
    return runner()
           ^^^^^^^^
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/scvi/train/_trainrunner.py", line 98, in __call__
    self.trainer.fit(self.training_plan, self.data_splitter)
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/scvi/train/_trainer.py", line 221, in fit
    super().fit(*args, **kwargs)
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/lightning/pytorch/trainer/trainer.py", line 544, in fit
    call._call_and_handle_interrupt(
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/lightning/pytorch/trainer/call.py", line 44, in _call_and_handle_interrupt
    return trainer_fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/lightning/pytorch/trainer/trainer.py", line 580, in _fit_impl
    self._run(model, ckpt_path=ckpt_path)
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/lightning/pytorch/trainer/trainer.py", line 989, in _run
    results = self._run_stage()
              ^^^^^^^^^^^^^^^^^
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/lightning/pytorch/trainer/trainer.py", line 1035, in _run_stage
    self.fit_loop.run()
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/lightning/pytorch/loops/fit_loop.py", line 202, in run
    self.advance()
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/lightning/pytorch/loops/fit_loop.py", line 359, in advance
    self.epoch_loop.run(self._data_fetcher)
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/lightning/pytorch/loops/training_epoch_loop.py", line 136, in run
    self.advance(data_fetcher)
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/lightning/pytorch/loops/training_epoch_loop.py", line 240, in advance
    batch_output = self.automatic_optimization.run(trainer.optimizers[0], batch_idx, kwargs)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/lightning/pytorch/loops/optimization/automatic.py", line 187, in run
    self._optimizer_step(batch_idx, closure)
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/lightning/pytorch/loops/optimization/automatic.py", line 265, in _optimizer_step
    call._call_lightning_module_hook(
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/lightning/pytorch/trainer/call.py", line 157, in _call_lightning_module_hook
    output = fn(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/lightning/pytorch/core/module.py", line 1282, in optimizer_step
    optimizer.step(closure=optimizer_closure)
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/lightning/pytorch/core/optimizer.py", line 151, in step
    step_output = self._strategy.optimizer_step(self._optimizer, closure, **kwargs)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/lightning/pytorch/strategies/strategy.py", line 230, in optimizer_step
    return self.precision_plugin.optimizer_step(optimizer, model=model, closure=closure, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/lightning/pytorch/plugins/precision/precision.py", line 117, in optimizer_step
    return optimizer.step(closure=closure, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/torch/optim/optimizer.py", line 373, in wrapper
    out = func(*args, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/torch/optim/optimizer.py", line 76, in _use_grad
    ret = func(self, *args, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/torch/optim/adamw.py", line 161, in step
    loss = closure()
           ^^^^^^^^^
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/lightning/pytorch/plugins/precision/precision.py", line 104, in _wrap_closure
    closure_result = closure()
                     ^^^^^^^^^
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/lightning/pytorch/loops/optimization/automatic.py", line 140, in __call__
    self._result = self.closure(*args, **kwargs)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/lightning/pytorch/loops/optimization/automatic.py", line 126, in closure
    step_output = self._step_fn()
                  ^^^^^^^^^^^^^^^
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/lightning/pytorch/loops/optimization/automatic.py", line 315, in _training_step
    training_step_output = call._call_strategy_hook(trainer, "training_step", *kwargs.values())
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/lightning/pytorch/trainer/call.py", line 309, in _call_strategy_hook
    output = fn(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/lightning/pytorch/strategies/strategy.py", line 382, in training_step
    return self.lightning_module.training_step(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/scvi/train/_trainingplans.py", line 350, in training_step
    _, _, scvi_loss = self.forward(batch, loss_kwargs=self.loss_kwargs)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/scvi/train/_trainingplans.py", line 284, in forward
    return self.module(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/scvi/module/base/_decorators.py", line 32, in auto_transfer_args
    return fn(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/scvi/module/base/_base_module.py", line 204, in forward
    return _generic_forward(
           ^^^^^^^^^^^^^^^^^
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/scvi/module/base/_base_module.py", line 744, in _generic_forward
    inference_outputs = module.inference(**inference_inputs, **inference_kwargs)
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/scvi/module/base/_decorators.py", line 32, in auto_transfer_args
    return fn(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/scvi/module/base/_base_module.py", line 308, in inference
    return self._regular_inference(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/scvi/module/base/_decorators.py", line 32, in auto_transfer_args
    return fn(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/scvi/module/_vae.py", line 337, in _regular_inference
    qz, z = self.z_encoder(encoder_input, batch_index, *categorical_input)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/scvi/nn/_base_components.py", line 288, in forward
    dist = Normal(q_m, q_v.sqrt())
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/torch/distributions/normal.py", line 56, in __init__
    super().__init__(batch_shape, validate_args=validate_args)
  File "/vf/users/CARD_singlecell/brain_atlas_wnn/.snakemake/conda/7eb17e13d8955d4ba42eb01fc9666aaf_/lib/python3.11/site-packages/torch/distributions/distribution.py", line 68, in __init__
    raise ValueError(
ValueError: Expected parameter loc (Tensor of shape (128, 16)) of distribution Normal(loc: torch.Size([128, 16]), scale: torch.Size([128, 16])) to satisfy the constraint Real(), but found invalid values:
tensor([[nan, nan, nan,  ..., nan, nan, nan],
        [nan, nan, nan,  ..., nan, nan, nan],
        [nan, nan, nan,  ..., nan, nan, nan],
        ...,
        [nan, nan, nan,  ..., nan, nan, nan],
        [nan, nan, nan,  ..., nan, nan, nan],
        [nan, nan, nan,  ..., nan, nan, nan]], device='cuda:0',
       grad_fn=<AddmmBackward0>)
```
@martinkim0 trying your suggestion now and will update on whether or not it works, thanks!

**Comment by shahrozeabbas** [comment]
@martinkim0 it seems that did not work unfortunately, same error :(

**Comment by canergen** [comment]
Possibly you have cells with very few ATAC reads. In my experience, this can lead to instabilities. Try filtering out cells with very low read counts (it's not happening below 100 reads).

**Comment by martinkim0** [comment]
Could you also try setting `scvi.settings.seed = 0` as well as passing in `train_size=1.0` to `train`?

**Comment by shahrozeabbas** [comment]
Hello,

I have gone back to the filtering step to make sure cells with low read counts are not included. Just changing that and running default settings resulted in the same error. 

I tried again using the previously suggested tips, but it doesn't seem to have worked. Using `train_size=1.0` has introduced a different error regarding the data loader. I assume that it expects some sort of split to take place. I have pasted my script below just in case. 

I wonder if it has something to do with the environment set up using `mamba`. 

```
import scvi
import muon
import torch

scvi.settings.seed = 0
torch.set_float32_matmul_precision('high')


adata = muon.read_anndata(filename=snakemake.input.muon_object, mod=snakemake.params.assay) # type: ignore


noise = ['nucleosome_percentile', 'TSS.percentile', 'frip']

scvi.external.POISSONVI.setup_anndata(adata, layer='counts', batch_key='sample', continuous_covariate_keys=noise) # type: ignore

model = scvi.external.POISSONVI(adata, n_latent=30, var_activation=torch.nn.functional.softplus) # type: ignore

model.train(accelerator='gpu', max_epochs=500, plan_kwargs={'lr_factor': 0.1}, train_size=1.0)


adata.obsm['X_poissonvi'] = model.get_latent_representation() # type: ignore


adata.write_h5ad(filename=snakemake.output.anndata, compression='gzip') # type: ignore


```

**Comment by martinkim0** [comment]
Hey @shahrozeabbas, so sorry you're still experiencing this issue. Is there a particular reason that you're using a non-default value of `n_latent`? Could you also check if not passing in `batch_key` and `continuous_covariate_keys` changes anything? Thanks.

**Comment by shahrozeabbas** [comment]
@martinkim0 thanks for the suggestion I will give it a try. It seems you are suggesting to just try the defaults and see if it works? Should I also try with or without `POISSONVI(var_activation=torch.nn.functional.softplus)`?

Admittedly, my understanding of the latent space is naive, but there isn't a particular reason for not using the default of `n_latent`. I've used a handful of the `scvi` models many times and typically try different values then compare the results. The dataset is 1.5 million cells and so my intuition was to increase the number. Have not explored the autotuner yet.

**Comment by shahrozeabbas** [comment]
@canergen Is the 100 reads lower limit potentially a standard for `poissonvi` or could it vary per dataset?

**Comment by shahrozeabbas** [comment]
Well after weeks of trying many, many things....I think I figured out what was causing the issue. I am calculating the 95% most commonly observed peaks across all samples and taking the union when calling `concat` to combine samples. I changed this to the intersection, basically an inner join of all peaks and `poissonvi` works as expected. Whatever the issue was when taking the union was not observed in only a few datasets and so everything would work fine. However, across all samples some artifact was appearing and creating a problem. 

If there are any thoughts or ideas as to what was specifically happening, would greatly appreciate your insight. Otherwise I can close this issue. Thanks again for your time!
